### ISSUE TITLE
Can I train a num_class is 26(>21) model use the SSD-512 VGG-based pre-trained checkpoints?

### ISSUE THREAD

**Issue by blacklong28** [balancap/SSD-Tensorflow#138]
Can I train a num_class is 36(>21) model use the SSD-512 VGG-based pre-trained checkpoints?
when I train a model which num_class is 16(<21), there is no problem . 
And I got a good result.
But when I add the number of the class ,I  got a error 
```
InvalidArgumentError (see above for traceback): Assign requires shapes of both tensors to match. lhs shape= [156] rhs shape= [126]
         [[Node: save_1/Assign_4 = Assign[T=DT_FLOAT, _class=["loc:@ssd_512_vgg/block10_box/conv_cls/biases"], use_locking=true, validate_shape=true, _device="/job:localhost/replica:0/task:0/cpu:0"](ssd_512_vgg/block10_box/conv_cls/biases, save_1/RestoreV2_4)]]
```
who can give me a hint or help ?

@balancap

**Comment by congjianting** [comment]
@blacklong617  did you fintune the model using vgg16.ckpt with pascalvoc 2007 ?

**Comment by blacklong28** [comment]
Yes, i did.
by use this command:
```
DATASET_DIR=./tfrecords
TRAIN_DIR=./log/
CHECKPOINT_PATH=./checkpoints/vgg_16.ckpt
python train_ssd_network.py \
    --train_dir=${TRAIN_DIR} \
    --dataset_dir=${DATASET_DIR} \
    --dataset_name=pascalvoc_2007 \
    --dataset_split_name=train \
    --model_name=ssd_300_vgg \
    --checkpoint_path=${CHECKPOINT_PATH} \
    --checkpoint_model_scope=vgg_16 \
    --checkpoint_exclude_scopes=ssd_300_vgg/conv6,ssd_300_vgg/conv7,ssd_300_vgg/block8,ssd_300_vgg/block9,ssd_300_vgg/block10,ssd_300_vgg/block11,ssd_300_vgg/block4_box,ssd_300_vgg/block7_box,ssd_300_vgg/block8_box,ssd_300_vgg/block9_box,ssd_300_vgg/block10_box,ssd_300_vgg/block11_box \
    --trainable_scopes=ssd_300_vgg/conv6,ssd_300_vgg/conv7,ssd_300_vgg/block8,ssd_300_vgg/block9,ssd_300_vgg/block10,ssd_300_vgg/block11,ssd_300_vgg/block4_box,ssd_300_vgg/block7_box,ssd_300_vgg/block8_box,ssd_300_vgg/block9_box,ssd_300_vgg/block10_box,ssd_300_vgg/block11_box \
    --save_summaries_secs=60 \
    --save_interval_secs=600 \
    --weight_decay=0.0005 \
    --optimizer=adam \
    --learning_rate=0.001 \
    --learning_rate_decay_factor=0.94 \
    --batch_size=32

```
and it can run .
but the loss cannot decrease.
and the result is bad.
so I give up use the vgg16.ckpt.

**Comment by congjianting** [comment]
@blacklong617  my result is very bad, the loss cannot decrease, if you do not finetune vgg_16.ckpt, how do you use this project for other dataset?

**Comment by blacklong28** [comment]
@congjianting 
When I train a model which num_class is 16(<21)，I use VGG_VOC0712_SSD_300x300_ft_iter_120000.ckpt or 512x512 ，the result is good.
But when I add the number of the class to 36(>21), I still use SSD-512 VGG-based pre-trained checkpoints, I got the error ：
InvalidArgumentError (see above for traceback): Assign requires shapes of both tensors to match. lhs shape= [156] rhs shape= [126]
         [[Node: save_1/Assign_4 = Assign[T=DT_FLOAT, _class=["loc:@ssd_512_vgg/block10_box/conv_cls/biases"], use_locking=true, validate_shape=true, _device="/job:localhost/replica:0/task:0/cpu:0"](ssd_512_vgg/block10_box/conv_cls/biases, save_1/RestoreV2_4)]]

I try to use vgg16.ckpt by：
DATASET_DIR=./tfrecords
TRAIN_DIR=./log/
CHECKPOINT_PATH=./checkpoints/vgg_16.ckpt
python train_ssd_network.py \
    --train_dir=${TRAIN_DIR} \
    --dataset_dir=${DATASET_DIR} \
    --dataset_name=pascalvoc_2007 \
    --dataset_split_name=train \
    --model_name=ssd_300_vgg \
    --checkpoint_path=${CHECKPOINT_PATH} \
    --checkpoint_model_scope=vgg_16 \
    --checkpoint_exclude_scopes=ssd_300_vgg/conv6,ssd_300_vgg/conv7,ssd_300_vgg/block8,ssd_300_vgg/block9,ssd_300_vgg/block10,ssd_300_vgg/block11,ssd_300_vgg/block4_box,ssd_300_vgg/block7_box,ssd_300_vgg/block8_box,ssd_300_vgg/block9_box,ssd_300_vgg/block10_box,ssd_300_vgg/block11_box \
    --trainable_scopes=ssd_300_vgg/conv6,ssd_300_vgg/conv7,ssd_300_vgg/block8,ssd_300_vgg/block9,ssd_300_vgg/block10,ssd_300_vgg/block11,ssd_300_vgg/block4_box,ssd_300_vgg/block7_box,ssd_300_vgg/block8_box,ssd_300_vgg/block9_box,ssd_300_vgg/block10_box,ssd_300_vgg/block11_box \
    --save_summaries_secs=60 \
    --save_interval_secs=600 \
    --weight_decay=0.0005 \
    --optimizer=adam \
    --learning_rate=0.001 \
    --learning_rate_decay_factor=0.94 \
    --batch_size=32

but the loss cannot decrease.
My doubt is when the class number is over 21,can we still use the pre-trained checkpoints which is train for a 21 class network.
The error I got is a Bug for this project? 
or I must use vgg_16.ckpt instead of VGG_VOC0712_SSD_512x512_ft_iter_120000.ckpt when the class number is over 21?

**Comment by congjianting** [comment]
@blacklong617 the class number changed, meaning that you need to exclude some layers to finetune.

**Comment by blacklong28** [comment]
@congjianting 
oh，your mean is the parameter:
--checkpoint_exclude_scopes
but can you tell me which layers should be exclude?

**Comment by cherryxiongyw** [comment]
@blacklong617  Do you solve the problem? I have the same problem, I train a new model and save it, so I get three files as:
`model.ckpt-N.data-00000-of-00001, model.ckpt-N.index, model.ckpt-N.meta` .
When I restore this checkpoint model to make prediction, I have the same error
`Assign requires shapes of both tensors to match. lhs shape= [126] rhs shape= [120]`
 @balancap 
Who can help me? Thanks!

**Comment by blacklong28** [comment]
@cherryxiongyw 
 I don't solve it.
I got the  error when I was training but not  making prediction.

**Comment by safehumeng** [comment]
@blacklong617 @cherryxiongyw checkpoint_exclude_scopes=ssd_300_vgg/conv6,ssd_300_vgg/conv7,ssd_300_vgg/block8,ssd_300_vgg/block9,ssd_300_vgg/block10,ssd_300_vgg/block11,ssd_300_vgg/block4_box,ssd_300_vgg/block7_box,ssd_300_vgg/block8_box,ssd_300_vgg/block9_box,ssd_300_vgg/block10_box,ssd_300_vgg/block11_box 
--trainable_scopes=ssd_300_vgg/conv6,ssd_300_vgg/conv7,ssd_300_vgg/block8,ssd_300_vgg/block9,ssd_300_vgg/block10,ssd_300_vgg/block11,ssd_300_vgg/block4_box,ssd_300_vgg/block7_box,ssd_300_vgg/block8_box,ssd_300_vgg/block9_box,ssd_300_vgg/block10_box,ssd_300_vgg/block11_box

**Comment by zzks** [comment]
i am trying the same thing but did not meet this problem, maybe you forgot to delete "antiques" in the checkpoint folder :)

**Comment by naresh98** [comment]
@zzks, 
can you please share your commnad to run/train?
Thanks in advance

**Comment by LJXLJXLJX** [comment]
@blacklong617 
hello , do you sovle the problem?
I am trying training a num_class >70 model ...

**Comment by vokhidovhusan** [comment]
Did anyone solve the problem?

**Comment by xiyuanzh** [comment]
@congjianting @blacklong617 @safehumeng 
Have you met the following problem when num_class is less than 21?
```
NotFoundError (see above for traceback): Restoring from checkpoint failed. This is most likely due to a Variable name or other graph key that is missing from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Key beta1_power not found in checkpoint
	 [[node save/RestoreV2 (defined at train_ssd_network.py:372)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_INT64, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]
```
My script is as below:
```
CUDA_VISIBLE_DEVICES=0 python3 train_ssd_network.py \
    --train_dir=./log_files/log_dog \
    --dataset_dir=./datasets/tfrecords \
    --dataset_name=pascalvoc_2007 \
    --dataset_split_name=train \
    --model_name=ssd_300_vgg \
    --checkpoint_path=./checkpoints/vgg_16.ckpt \
    --checkpoint_model_scope=vgg_16 \
    --checkpoint_exclude_scopes=ssd_300_vgg/conv6,ssd_300_vgg/conv7,ssd_300_vgg/block8,ssd_300_vgg/block9,ssd_300_vgg/block10,ssd_300_vgg/block11,ssd_300_vgg/block4_box,ssd_300_vgg/block7_box,ssd_300_vgg/block8_box,ssd_300_vgg/block9_box,ssd_300_vgg/block10_box,ssd_300_vgg/block11_box \
    --trainable_scopes=ssd_300_vgg/conv6,ssd_300_vgg/conv7,ssd_300_vgg/block8,ssd_300_vgg/block9,ssd_300_vgg/block10,ssd_300_vgg/block11,ssd_300_vgg/block4_box,ssd_300_vgg/block7_box,ssd_300_vgg/block8_box,ssd_300_vgg/block9_box,ssd_300_vgg/block10_box,ssd_300_vgg/block11_box \
    --save_summaries_secs=60 \
    --save_interval_secs=600 \
    --weight_decay=0.0005 \
    --optimizer=adam \
    --learning_rate=0.001 \
    --batch_size=32 \
```
Thanks a lot for helping!!!

**Comment by zyz111** [comment]
@blacklong617 
hello,
When I train a model which num_class is 8(<21)，I use ssd_ 512_vgg pretrain model ，the result is bad. So, can you show your script？ Thanks a lot for your help!

**Comment by blacklong28** [comment]
@zyz111 
Hi，time is too long, I forgot the script, you can refer to the author's README.

**Comment by zyz111** [comment]
@blacklong617 
Thank you

**Comment by JulioZhao1997** [comment]
you can exclude some box layers as follows:
checkpoint_exclude_scopes=ssd_300_vgg/block11_box,ssd_300_vgg/block11_box/conv_cls/weights,ssd_300_vgg/block10_box,ssd_300_vgg/block9_box,ssd_300_vgg/block8_box,ssd_300_vgg/block7_box,ssd_300_vgg/block4_box \
using this command to load pretrained model but i dont't know its effects on training performance

**Comment by zyz111** [comment]
thank you very much





---Original---
